content/blog/2019-11-13-efficient-path-prof/index.md

```diff
@@ -0,0 +1,205 @@
+
+Finally, the algorithm inserts instrumentation according to the `Inc` mapping. For every edge in the original CFG with a non-zero `Inc` value, the instruction `r += Inc(e)` is inserted by creating a new basic block between the source and destination of the edge with that instruction. The block is then inserted into the CFG by rerouting the source's successor to the new block and adding a jump at the end of the new block to the destination block. The entry block adds an instruction to initialize `r=0` and the exit block adds an instruction `path_table[r++]`. Back edges are bit more interesting. For back edges we insert the instruction sequence:
+
+```
+r += backedge_inc
+path_table[r++]
+r = backedge_reset
+```
+
+`backedge_inc` is set to `Inc((entry,backedge.dest))`, which is the `Inc` value of the edge from the entry block to the head of the loop containing the back edge. `backedge_reset` is set to `Inc((backedge.source, exit))`, which is the `Inc` value of the edge from the bottom of the loop to the exit block. This was somewhat underspecified in the paper but I believe it makes sense based on their descriptions of what the two inserted edges for each back edge represent. This is the instrumented CFG of the example we have been looking at:
+
+<img src="simple_inst.png" style="width: 50%">
+
+We can verify that our path mappings are preserved by figuring out the value of `r` at the end of every path before we do `path_table[r]++`:
+
+| Path | r  |
+|:-:|:-:|
+| A,B,D  | 0 |
+| A,B,C  | 1 |
+| B,C  | 3 |
+| B,D  | 2 |
+
+Also, note that this matches the `Val` sum for each path computed above.
+
+### Differences between the paper
+
+First, the paper seems to be a bit misleading when it comes to instrumenting back edges. It seems to imply that whenever you have a back edge that you should always insert the instructions `path_table[r]++; r = 0`. However, as the above simple example showed and the edge cases below will show, it actually needs to be a bit more complex than just increment and reset `r` to 0. The edge cases analyzed below will hopefully show that the following code snippet is actually what back edges should be instrumented with:
+
+```
+r += backedge_inc
+path_table[r++]
+r = backedge_reset
+```
+
+I also diverged from the paper when constructing the maximum spanning tree. The authors used existing edge profiling information to estimate the cost of instrumenting an edge. Because instrumentation is only placed on the chords of the maximum spanning tree, those will be the edges least likely to be taken. When I construct the maximum spanning tree, I don't assume any prior profiling information and assume that each edge is equally likely to be taken. This produces an instrumentation with minimal edges but not necessarily the edges that are taken the least. I felt like this was an acceptable compromise in order to keep this project fairly self contained and not rely on importing information from an edge profiler.
```

Comment:
Very much so. It's kind of a strange thing in this paper that they assume they get to use an edge profile to decide how to do their path profiling. You can undoubtedly do better that way, but the use case is kind of dubious.